### PR TITLE
mtmd : add support for Qwen2-Audio and SeaLLM-Audio

### PR DESCRIPTION
**Pull request by ngxson**
This PR adds support for:
- https://huggingface.co/Qwen/Qwen2-Audio-7B-Instruct
- https://huggingface.co/SeaLLMs/SeaLLMs-Audio-7B

> [!IMPORTANT]
>
> ## This model produces very poor result even on text-only tasks.

It is **not** practically usable, often hallucinates after a short amount of generated text (even on with f16 precision). However, I'm attempting to get it supported, so I can try out Qwen2.5-Omni next (only audio+image input, no audio output)

Due to the low quality, I think I'll skip uploading quant this time. Maybe I'll need to put up a notice in the `multimodal.md` (edit: added)

Also ref the newly created discussion: https://github.com/ggml-org/llama.cpp/discussions/13759

---

Some researches which made me conclude that the text model has problem:

<details><summary>It consistently goes bi-language (test both Qwen+SeaLLM, F16 model, same problem)</summary>

image

image

</details>

<details><summary>The official demo (non-llama.cpp) hallucinates content of the audio. I uploaded Martin Luther King Jr. "I have a dream" but it replies with nonsense music theory</summary>

Official demo: https://huggingface.co/spaces/Qwen/Qwen2-Audio-Instruct-Demo

Why I said "nonsense music theory"? Try playing it on the piano, it doesn't sound right 😂 

Screenshot 2025-05-24 at 21 29 58

</details>

<details><summary>I checked the encoder's logits from transformers and my implementation, they all matched up, so it's very likely that the text model has problem</summary>

image

</details>

<details><summary><del>The good thing is that, SeaLLM performs well on asian languages.</del> But ultravox does an even better job</summary>

Input audio: a recording about [Hue](https://en.wikipedia.org/wiki/Hu%E1%BA%BF)

SeaLLM (response correctly, but misses a lot of information):

image

Ultravox (very detailed and accurate response):

image

</details>

### CODE REVIEW

File: tools/mtmd/clip.cpp [hunk 1]
```diff
@@ -3450,6 +3485,10 @@ int clip_n_output_tokens(const struct clip_ctx * ctx, struct clip_image_f32 * im
         const int proj_stack_factor = ctx->vision_model.hparams.proj_stack_factor;
         const int n_len = CLIP_ALIGN(img->nx, proj_stack_factor);
         n_patches = n_len / proj_stack_factor / 2;
+    } else if (ctx->proj_type == PROJECTOR_TYPE_QWEN2A) {
```
Comment (ggerganov) [inline tools/mtmd/clip.cpp:3488]: nit: this should become a `switch` (no need to change in this PR)

File: tools/mtmd/clip.cpp [hunk 2]
```diff
@@ -1655,6 +1667,17 @@ struct clip_graph {
             inpL = cur;
         }
 
+        // TODO @ngxson : find a way to move this outside
+        if (ctx->proj_type == PROJECTOR_TYPE_QWEN2A) {
+            ggml_tensor * cur = inpL;
+            cur = ggml_transpose(ctx0, cur);
+            cur = ggml_cast(ctx0, cur, GGML_TYPE_F32);
```
Comment (ggerganov) [inline tools/mtmd/clip.cpp:1674]: Any reason to prefer `ggml_cast` here over `ggml_cont`?

**Review by ggerganov — APPROVED** [review]

File: tools/mtmd/clip.cpp (hunk 2 above)
Comment (ngxson) [reply tools/mtmd/clip.cpp]: I had some problem with `ggml_compute_forward_pool_1d` and when looking into the source code, I mistakenly thought that it only supports F16. Changed to `ggml_cont` in https://github.com/ggml-org/llama.cpp/pull/13760/commits/e53a0dc43f0869388e6169547abfc112184ab84d

**Comment by zhouwg** [comment]
ngxson, could you add support for ByteDance's latest multimodal BAGEL(https://huggingface.co/ByteDance-Seed/BAGEL-7B-MoT) in your great/excellent MTMD subsystem?

as well known, the DeepSeek and Alibaba Qwen and Bytedance Seekd and Tencent Hunyuan are the 4 most powerful/creative top AI lab/team in China.

**Comment by henfiber** [comment]
> so I can try out Qwen2.5-Omni next (only audio+image input, no audio output)

another interesting one is MiniCPM-o (2.6) which you have already added vision support for in MTMD. It seems competitive with Ultravox in [benchmarks](https://github.com/MatthewCYM/VoiceBench?tab=readme-ov-file#leaderboard). The benefit of these two (Qwen2.5-omni & MiniCPM-o) is that you can have one model loaded (with two different mmproj files?) for both image and audio, so you can spare some VRAM.   

MiniCPM-o has a benefit in image tokenization speed (much faster than Qwen2.5-VL-7b), so it is usable also on CPUs. Maybe that's true with their audio processing as well (no idea). The license is more restrictive though.

Needless to say, your work in image & audio support is much appreciated.

**Comment by foldl** [comment]
## Does Qwen2-Audio perform poorly?

### Audio

I came across this issue from doc of `mtmd`. I am curious about what's going on. 

Since @ngxson had not shared the 30 second audio clip, I extracted part of the recording from [here](https://archive.org/details/MLKDream) (see attachment) and tested it with [chatllm.cpp](https://github.com/foldl/chatllm.cpp). I think the result is not poor at all. The model is `q8_0` quantized.

```
main.exe -m quantized\qwen2-audio-7b.bin -ngl all --multimedia-file-tags {{ }} -i
    ________          __  __    __    __  ___ 
   / ____/ /_  ____ _/ /_/ /   / /   /  |/  /_________  ____  
  / /   / __ \/ __ `/ __/ /   / /   / /|_/ // ___/ __ \/ __ \ 
 / /___/ / / / /_/ / /_/ /___/ /___/ /  / // /__/ /_/ / /_/ / 
 \____/_/ /_/\__,_/\__/_____/_____/_/  /_(_)___/ .___/ .___/  
You are served by Qwen2-Audio,                /_/   /_/       
with 7754878976 (7.8B) parameters.

You  > what is this? {{audio:.../MLKDream_64kb.wav}}
A.I. > This is the transcript of a speech given by an individual at an event.

You  > transcribe what he said
A.I. > The speech said was: 'Fivescore years ago a great American in whose symbolic shadow we stand today signed the Emancipation Proclamation.'
```

P.S. I think this well-known recording is probably in the training set.

[MLKDream_64kb.zip](https://github.com/user-attachments/files/20874363/MLKDream_64kb.zip)

### Text-only

I failed to re-produce the bi-language issue. Tested for two times, one with random sampling, one using greedy sampling. `q8_0` quantization.

```
You  > write a very technical blog post about LLM
A.I. > Long Short-Term Memory (LSTM) is a type of recurrent neural network (RNN) architecture that is particularly well-suited for modeling sequential data. It was first introduced in 2009 by researchers at the University of Toronto, and has since become one of the most widely-used and successful architectures for a variety of natural language processing tasks.
One of the key advantages of LLMs is that they are able to effectively model long-term dependencies in sequential data. This is in contrast to traditional RNNs, which struggle with this task due to the vanishing gradient problem. The vanishing gradient problem occurs when the gradients of the loss function with respect to the model parameters become very small, making it difficult for the model to learn long-term dependencies.
LLMs address this problem by using a technique called "gated" units, which allow the model to selectively remember or forget information from previous time steps. This is achieved through the use of two gates: the "input gate" and the "output gate", which control the flow of information into and out of the cell state of the LSTM.
In addition to the gated units, LLMs also use a mechanism called "cell state" to maintain information over longer periods of time. The cell state is a hidden state that is passed from one time step to the next, and it is updated based on the input at the current time step and the previous cell state.
LLMs have been highly successful in a wide range of natural language processing tasks, including language modeling, machine translation, sentiment analysis, and question answering. They have also been used in a variety of other applications, such as speech recognition and computer vision.
Overall, LLMs are a powerful and versatile architecture for modeling sequential data, and they have had a significant impact on the field of natural language processing.
```

Greedy:

```
You  > write a very technical blog post about LLM
A.I. > Long Short-Term Memory (LSTM) is a type of recurrent neural network (RNN) architecture that is widely used for natural language processing tasks such as language modeling, machine translation, and sentiment analysis. It is particularly effective at handling long-term dependencies in sequential data, which is a common problem in natural language processing.
One of the key features of LLMs is their ability to maintain long-term dependencies by using a mechanism called "gates" that control the flow of information through the network. This allows the network to selectively remember or forget information from previous time steps, which is particularly useful for tasks that require the network to maintain information over long periods of time.
Another important feature of LLMs is their ability to handle variable-length sequences, which is a common problem in natural language processing tasks. This is achieved through the use of a dynamic memory cell that can store information of different lengths, allowing the network to handle variable-length sequences without the need for explicit padding or truncation.
In addition to these key features, LLMs also have a number of other advantages over other RNN architectures. For example, they are more computationally efficient, as they do not require the network to recompute the same values multiple times. They are also more robust to vanishing gradients, which can be a problem in traditional RNNs when training on long sequences.
Overall, LLMs are a powerful and versatile architecture that have proven to be highly effective for a wide range of natural language processing tasks. They are a key component of many state-of-the-art models in this field, and their ability to handle long-term dependencies and variable-length sequences makes them an essential tool for researchers and practitioners alike.
```
